packages/client-utils/src/vstorage-client.js

```diff
@@ -0,0 +1,489 @@
+ * @returns {Topic<T>}
+ */
+const makeTopic = ({ queryClient }, path, options) => ({
+  latest: async height => {
+    if (height && height < MINIMUM_HEIGHT)
+      throw createError(
+        `${INVALID_HEIGHT_ERROR_MESSAGE} ${height}`,
+        height,
+        PATHS.DATA,
+        path,
+      );
+
+    const response = await queryClient.queryAbci(path, {
+      height:
+        height && height > MINIMUM_HEIGHT && options?.fetchDataAtPreviousHeight
+          ? height - 1n
+          : height,
+      kind: PATHS.DATA,
+    });
+    const parsedValue = parseValue(response.value);
+
+    if (options?.assertCell && !isStreamCell(parsedValue))
+      throw Error(`Expected a stream cell, got ${quote(parsedValue)}`);
+    else if (options?.assertDataString && !isDataString(parsedValue))
+      throw Error(`Expected a data string, got ${quote(parsedValue)}`);
+
+    return /** @type {T} */ (parsedValue);
+  },
+});
+
+/**
+ * @param {{fetch: typeof window.fetch}} powers
+ * @param {MinimalNetworkConfig} config
+ */
+export const makeVStorageClient = ({ fetch }, config) => {
+  if (!config.rpcAddrs.length)
+    return Fail`${INVALID_RPC_ADDRESS_ERROR_MESSAGE} ${quote(config)}`;
+
+  const queryClient = makeQueryClient({ fetch }, config);
+
+  /**
+   * @template {any} T
+   * @param {string} path
+   * @param {Partial<{ compat: boolean }>} [options]
+   */
+  const fromText = (path, options) =>
+    /** @type {StreamTopic<T>} */ (
+      makeStreamTopic({ queryClient }, path, options)
+    );
+
+  const vStorageClient = {
+    fromText,
+    /**
+     * @param {string} path
+     */
+    fromTextBlock: path =>
+      /** @type {Topic<string>} */ (
+        makeTopic({ queryClient }, path, {
+          assertDataString: true,
+          fetchDataAtPreviousHeight: true,
+        })
+      ),
+    /**
+     * @param {string} path
+     */
+    keys: async path => {
+      const response = await queryClient.queryAbci(path, {
+        kind: PATHS.CHILDREN,
+      });
+      return response.children;
+    },
+    /**
+     * @deprecated
+     * @template {any} T
+     * @param {string} path
+     * @param {Height} height
+     */
+    readAt: (path, height) =>
+      /** @type {Promise<StreamTopic<T>>} */ (fromText(path).latest(height)),
+    /**
+     * @deprecated
+     * @template {any} T
+     * @param {string} path
+     * @param {Height} [minimumHeight]
+     */
+    readFully: async (path, minimumHeight) => {
+      /** @type {Array<Update<T>>} */
+      const values = [];
+      for await (const value of /** @type {AsyncIterable<Update<T>>} */ (
+        fromText(path).reverseIterate(undefined, minimumHeight)
+      ))
+        values.push(value);
+      return values;
+    },
+    /**
+     * @deprecated
+     * @template {any} T
+     * @param {string} path
+     */
+    readLatest: path => /** @type {Promise<T>} */ (fromText(path).latest()),
```

Comment:
Since this module is a rewrite of `vstorage.js`, we do not need to construct the deprecated methods at all. We just need these deprecation notes to be added to the existing implementation, recommending that they import this module instead and use the modern interface.